### ISSUE TITLE
Extracting component to its own npm package

### ISSUE THREAD

**Issue by pavlin-policar** [mxstbr/react-boilerplate#955]
I had been working on a project for a while, and decided that one of the components was generalised enough to go into its own package, that I would then include in my main project. The component wasn't done, but it made more sense that it be separate from the main project. I tried to include the package into my main project using `npm link`. The package is written with react and es6, but has no need for HRM because its just a library. Because the HRM babel preset was not present in the package, I got this error while trying to run `npm start` in my main project:

```
Module build failed: Error: Couldn't find preset "react-hmre" relative to directory ".../local_package/lib"
```

This is not an issue with this boilerplate, but with the babel loader and webpack, there are many issues open regarding this. They also mention a workaround, which worked for me (kind of).

By changing line 61 in `internals/webpack/webpack.dev.babel.js`, from `presets: ['react-hmre'],` to `presets: [require.resolve('babel-preset-react-hmre')],` I am now able to work on both my local npm package and my main project (which uses this boilerplate) with the main project running with hot reloading for both the main project and package. I also had to add my local project to the `dllPlugin::exclude` packages. Obviously this only works for the HRM preset, and the same would have to be done to other presets as well, but this worked for me.

The key thing here is that now I can write my local package in es6 without having to build the package whenever I make a change, so in the `package.json` of my local package I can have `"main": "src/index.js"` instead of `"main": "lib/index.js"`.

This fails when trying to build the package due to previously mentioned issues in webpack. Also, the main issue for me was ease of development.

I think this should be documented somewhere (if I missed this in the docs on this, I apologise), since one of the most compelling points of the self contained containers/components to me was that you could easily extract them from the project into their own package.

My solution isn't perfect, but I would love to get input on how this could be improved, or what a better way of working on local packages in parallel would be.

**Comment by mxstbr** [comment]
Absolutely agree, this warrants a discussion! I don't have time to dive in at the moment, but I would love for you to submit a PR adding docs and fixing bugs to make this easier to do!

**Comment by pavlin-policar** [comment]
Sure thing, I'll do what I can!

**Comment by pavlin-policar** [comment]
I've created a PR #979 that adds some the documentation and fix for what I've described above.

There are still a couple of issues and I don't have much experience with webpack to know what causes some of them. The development process using my fix is very nice, webpack updates the bundle each time the linked package is changed.

A minor annoyance is that each time you want to publish the separate package to the npm registry, you've got to make sure to change the `src/index.js` in the `package.json` to `lib/index.js`. This can get very annoying. One solution would be to write a simple `sed` script that would change it before publishing, but that's limited to unix (I don't believe Windows has `sed`), so that might be able to be improved.

The big thing is that the build process fails when trying to compile a linked module. I'm actually not sure what's the cause for this, as far as I can tell, it's either webpack or babel. The best solution I've come up with is to remove the symlink from `node_modules` and do a quick `npm install` to fetch the latest published version of the package. This is, again, very annoying, but I don't run the build command very often. This should probably be improved, because I imagine if you build often, this would become more than a little bit annoying, but I'm not sure how I would go about fixing this.

Any suggestions or feedback would be welcome.

**Comment by bugkiwi** [comment]
`npm link` is a good choice, but this solution can't work well.
Look this [babel/babel-loader#149](https://github.com/babel/babel-loader/issues/149), [webpack/webpack#1866](https://github.com/webpack/webpack/issues/1866).

It's seem only `babel-core@6.1.21` can work, but when I try it in [react-boilerplate@3.1.0](https://github.com/mxstbr/react-boilerplate/releases/tag/v3.1.0), it will lead to more errors.

Now I have try serveral methods,like give a new presets in `package.json` or `webpack.dev.babel.js` but my third package still can't find `react`

>Module not found: Error: Can't resolve 'react' in /Users/xxx/third/package/index.js;

**Comment by pavlin-policar** [comment]
Yes, I was wondering why I wasn't getting this error but, this should be easily fixable, as it's the same problem I ran into with HRM. I've seen the issues you linked, and the fix should apply to this as well. It's just a matter of finding where to stick the `require.resolve` in the config files.

This may be tricky because some of the presets and plugins are specified in the `package.json` from which (as I understand), babel reads directly, so that may be an issue. Perhaps moving those settings to the `webpack.dev.babel.js` with `.map(require.resolve)` may fix the issue?

**Comment by bugkiwi** [comment]
@pavlin-policar Thanks for reply;

My `webpack.dev.babel.js` looks like this:

```
  babelQuery: {
    // presets: ['react-hmre'],
    // http://stackoverflow.com/questions/34574403/how-to-set-resolve-for-babel-loader-presets/
    presets: [
      'babel-preset-es2015',
      'babel-preset-stage-0',
      'babel-preset-react',
      'babel-preset-react-hmre'
    ].map(require.resolve),
    //plugins: ['babel-plugin-transform-async-to-generator'].map(require.resolve)
  },

```
and It will lead to `Module not found: Error: Can't resolve'react' in` and when i remove my third package and just run a base example, lead to other error in brower:

`store.js?fa02:12 Uncaught TypeError: (0 , _reduxSaga2.default) is not a function`

**I don't install any packages in my third package, `node_modules` is empty.**

**Comment by pavlin-policar** [comment]
@bugkiwi This last bug doesn't seem related to this particular issue. I know there are some issues with the current release where the dependencies are outdated and cause errors, but those are easily fixable.

I have, however, played around with removing some of the babel presets to the `webpack.base.js`, since I don't think there's a way to change how webpack resolves relative imports from there.

This function
```js
const resolvePreset = (preset) => {
  if (Array.isArray(preset)) {
    const [p, ...opts] = preset;
    return [require.resolve(`babel-preset-${p}`), ...opts];
  }
  return require.resolve(`babel-preset-${preset}`);
}
```
along with changing the babel options in `webpack.base.js` to
```js
    {
      test: /\.js$/, // Transform all .js files required somewhere with Babel
      loader: 'babel',
      exclude: /node_modules/,
      query: {
        presets: [
          ['es2015', { modules: false }],
          'react',
          'stage-0',
          ...(options.babelQuery ? options.babelQuery.presets : [])
        ].map(resolvePreset),
        plugins: [],
      }
```
I removed all the changes I made in `webpack.dev.js` and removed the options from `package.json` and this worked. I haven't tested it for plugins, though I imagine it should be relatively straightforward.

**Comment by bugkiwi** [comment]
@pavlin-policar Yeah, I had try give a simple pure js without any dependencies in my third lib, everything worked well.

I think if I give my third lib a completely package, and compile them, everything should work well; but i don't have time to do that, and now i just use a copy on my apps.When I have some free time, I will try that again;

**Comment by bugkiwi** [comment]
A few days ago I tried it, all work well.
Before there are some problems because I cited the wrong lead, and I'm very sorry.
Thank you @pavlin-policar :)

**Comment by sedubois** [comment]
I'm not very familiar with Babel, but why are some things configured in `package.json` while one thing is configured in [`webpack.dev.babel.js#L45`](https://github.com/mxstbr/react-boilerplate/blob/master/internals/webpack/webpack.dev.babel.js#L45)?

**Comment by Dattaya** [comment]
> I'm not very familiar with Babel, but why are most things configured in package.json with the exception of one thing configured in webpack.dev.babel.js#L45?

If someone wants to transpile a server with babel also, and today it happed: #1170, hot reloading won't work in that environment.

**Comment by mxstbr** [comment]
> I'm not very familiar with Babel, but why are most things configured in package.json with the exception of one thing configured in webpack.dev.babel.js#L45?

That's a great point actually. We should move that back to the `package.json` in `dev`!

**Comment by Dattaya** [comment]
I had to do that in a universal app, but maybe here it's fine as `module.hot` is only used on the client?

**Comment by mxstbr** [comment]
Documented by the amazing @pavlin-policar in #979!

**Comment by hcschuetz** [comment]
I've also extracted a component to a separate package and used a slightly different approach to preserve hot reloading for the sources of that component:
- I'm using webpack's [resolve.alias](https://webpack.js.org/configuration/resolve/#resolve-alias) rather than `npm link` for referring to the component.  (This was not for a deep reason but just because I was not aware of `npm link`.)
- The idea was to define the alias in `webpack.dev.babel.js`, whereas `webpack.prod.babel.js` does not contain the alias.  The alias points to the component's `src` directory.
  - To make this work I had to change `webpack.base.babel.js` to merge `options.resolve` into its own `resolve` parameter.  (The standard code from `react-boilerplate` ignored `options.resolve`.)
    - BTW: `webpack.base.babel.js` implements the merging of the parameters from the specific configuration with its own base-configuration parameters.  In this situation it would have been more comfortable if the merging were implemented in `webpack.(dev|test|prod).babel.js`. `webpack.base.babel.js` should just export its parameters as a data object.  (But there might be other situations where the current behavior is more comfortable.)
  - I also got the error message about the `react-hmre` preset as in the original post of this issue.  (I actually found the current issue by googling for the error message.)  @pavlin-policar's solution with `[require.resolve('babel-preset-react-hmre')]` worked for me as well.  Thanks a lot!
- I also added the alias to `webpack.test.babel.js` to avoid eslint errors about unresolvable imports. (In `package.json` eslint is configured to use that file for module resolving.)

Hot reloading from the extracted component works nicely.  Also there is no need to tweak the "main" parameter of the component's `package.json` since my alias points directly to the source directory.  And finally the approach does not involve my (and my colleague's) global npm repository.

However, my experience is limited for now:
- My extracted component has only a single dependency.  (And in particular it does not depend on React.)  So I do not yet know how well the component's dependencies interact with the main app's dependencies.
- I'm not yet actually creating prod builds.

Maybe this approach is useful for somebody.  If you are interested I can go into more detail or report about how it's going when I'll extract more complex components.
